broker-daemon/block-order-worker/index.js

```diff
@@ -402,55 +416,69 @@ class BlockOrderWorker extends EventEmitter {
       throw new Error(`No engine available for ${counterSymbol}`)
     }
 
-    const promisedFills = orders.map(async (order) => {
+    const ordersByOrderId = await Promise.all(orders.map((order) => {
+      const range = {
+        gte: order.orderId,
+        lte: order.orderId
+      }
+      return getRecords(this.ordersByOrderId, Order.fromStorage.bind(Order), this.ordersByOrderId.range(range))
+    }))
+
+    const ownOrderIds = ordersByOrderId.filter(order => order && order.length > 0).map(([order]) => order.orderId)
+
+    const promisedFills = []
+    for (let order of orders) {
       const depthRemaining = targetDepth.minus(currentDepth)
 
       // if we have already reached our target depth, create no further fills
       if (depthRemaining.lte(0)) {
         return
       }
 
+      if (ownOrderIds.includes(order.orderId)) {
+        throw new Error(`Cannot fill own order ${order.orderId}`)
+      }
+
       // Take the smaller of the remaining desired depth or the base amount of the order
       const fillAmount = depthRemaining.gt(order.baseAmount) ? order.baseAmount : depthRemaining.toString()
 
       // track our current depth so we know what to fill on the next order
       currentDepth = currentDepth.plus(fillAmount)
 
-      const fsm = await FillStateMachine.create(
-        {
-          relayer,
-          engines,
-          logger,
-          store
-        },
-        blockOrder.id,
-        order,
-        { fillAmount }
-      )
-
-      // We are hooking into the execute lifecycle event of a fill state machine to trigger
-      // the completion of a blockorder
-      fsm.once('execute', () => {
-        this.checkBlockOrderCompletion(blockOrder.id)
-          .catch(e => {
-            this.logger.error(`BlockOrder failed to be completed from fill`, { id: blockOrder.id, error: e.stack })
-          })
-          .then(() => fsm.removeAllListeners())
-      })
-
-      // We are hooking into the reject lifecycle event of a fill state machine to trigger
-      // the failure of a blockorder
-      fsm.once('reject', () => {
-        this.failBlockOrder(blockOrder.id, fsm.fill.error)
-          .catch(e => {
-            this.logger.error(`BlockOrder failed on setting a failed status from fill`, { id: blockOrder.id, error: e.stack })
+      promisedFills.push(
+        FillStateMachine.create(
+          {
+            relayer,
+            engines,
+            logger,
+            store
+          },
+          blockOrder.id,
+          order,
+          { fillAmount }
+        ).then((fsm) => {
```

Comment:
is it worth it to move this function somewhere else? Like a dedicated `applyFsmListeners` function? Feels a bit messy in the loop